src/main/java/com/google/sps/servlets/InfoCardServlet.java

```diff
@@ -55,115 +50,30 @@
    */
   @Override
   public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String address = request.getParameter("address");
-    if (address == null) {
-      response.setContentType("text/html");
-      response
-          .getWriter()
-          .println("No address in the query URL, please check why this is the case.");
-      response.setStatus(400);
-      return;
-    }
+    Optional<String> optionalAddress = ServletUtils.getRequestParam(request, response, "address");
+    Optional<String> optionalElectionId =
+        ServletUtils.getRequestParam(request, response, "electionId");
 
-    String electionId = request.getParameter("electionId");
-    if (electionId == null) {
-      response.setContentType("text/html");
-      response
-          .getWriter()
-          .println("No election ID in the query URL, please check why this is the case.");
-      response.setStatus(400);
+    if (!optionalAddress.isPresent() || !optionalElectionId.isPresent()) {
       return;
     }
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Entity chosenElection;
+    Optional<Election> optionalElection =
+        ServletUtils.findElectionInDatastore(datastore, optionalElectionId.get());
 
-    Optional<Entity> foundElection = ServletUtils.findElectionInDatastore(datastore, electionId);
-    if (!foundElection.isPresent()) {
-      response.setContentType("text/html");
-      response
-          .getWriter()
-          .println(String.format("Could not find election with ID %s in Datastore.", electionId));
-      response.setStatus(400);
+    if (!optionalElection.isPresent() || (optionalElection.get().isPopulatedByVoterInfoQuery())) {
```

Comment:
Consider logging this case for debugging? Or even returning an error if you think it's "fatal".